### ISSUE TITLE
manual_async_fn false positive

### ISSUE THREAD

**Issue by Seeker14491** [rust-lang/rust-clippy#5765]
The following code triggers the `manual_async_fn` lint:

```rust
use std::future::Future;

struct A;

impl A {
    fn f(&self) -> impl Future<Output = ()> {
        async {}
    }
}

fn main() {
    let _future = {
        let a = A;
        a.f()
    };
}
```

However, applying the suggestion of using an `async fn` causes the code to stop compiling:

```
error[E0597]: `a` does not live long enough
  --> src\main.rs:14:9
   |
12 |     let _future = {
   |         ------- borrow later stored here
13 |         let a = A;
14 |         a.f()
   |         ^ borrowed value does not live long enough
15 |     };
   |     - `a` dropped here while still borrowed
```

This is due to `async fn` implicitly capturing the lifetime of `&self` in the returned future, while the `impl Future` syntax does not.

### Meta

- `cargo clippy -V`: clippy 0.0.212 (50fc24d8a 2020-06-25)
- `rustc -Vv`:
  ```
  rustc 1.46.0-nightly (50fc24d8a 2020-06-25)
  binary: rustc
  commit-hash: 50fc24d8a172a853b5dfe40702d6550e3b8562ba
  commit-date: 2020-06-25
  host: x86_64-pc-windows-msvc
  release: 1.46.0-nightly
  LLVM version: 10.0
  ```

**Comment by ebroto** [comment]
To be consistent with `async fn`, we should check that the returned future captures all the input lifetimes, or that there are no input lifetimes. See [here](https://github.com/rust-lang/rfcs/blob/master/text/2394-async_await.md#lifetime-capture-in-the-anonymous-future).

I will work on fixing this in the coming days.

**Comment by ebroto** [comment]
So holidays and other stuff came and I could not dedicate a lot of time to clippy until now. Sorry for the delay!
